---
title: "Single cell transcriptomes of nk mono cells with HIV 2023"
author: "Burnet Bioinformatics group"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---

## Introduction

There are 3 samples.

1. 1-17032023-GEX

2. 2-09032023-GEX

3. 3-13032023-GEX

I had to make a new custom reference genome with the HIV sequence.

```{r,libs}

suppressPackageStartupMessages({
  library("plyr")
  library("Seurat")
  library("hdf5r")
  library("SingleCellExperiment")
  library("parallel")
  library("stringi")
  library("beeswarm")
  library("muscat")
  library("DESeq2")
  library("mitch")
  library("limma")
  library("kableExtra")
  library("gplots")

  library("network")
  library("eulerr")
  library("dplyr")
  library("reshape2")
})

knitr::opts_chunk$set(dev = 'svg') # set output device to svg

```

```{r,functions}

map2color<-function(x,pal,limits=NULL){
    if(is.null(limits)) limits=range(x)
    pal[findInterval(x,seq(limits[1],limits[2],length.out=length(pal)+1), all.inside=TRUE)]
}

gs2net <- function(gset,em,colfunc=colorRampPalette(c("blue", "white","red"))(n=100)){
  gset <- gset[which(lapply(gset,length)>0)]
  mydf <- bind_rows(lapply(gset, as.data.frame.list))
  rownames(mydf) <- names(gset)
  j <- apply(mydf,1,function(x) {
    apply(mydf,1,function(y) {
      length(intersect(x,y) ) / length(union(x,y))
    })
  })
  j[lower.tri(j)] <- NA
  j[lower.tri(j,diag=TRUE)] <- 0
  jl <- melt(j)
  jl <- jl[which(jl$Var1 != jl$Var2),]
  jl <- jl[which(jl$value != 1),]
  jl <- jl[order(-jl$value),]
  jl <- head(jl,length(gset)*2)
  jl$edgeSize = with(jl, jl$value/sum(jl$value))
  lengths <- unlist(lapply(gset,length))
  lengths <- sqrt(lengths/sum(lengths)*100)
  jl$vertexsize <- lengths[match(as.character(jl$Var1),names(lengths))]
  jl2 <- apply(jl[,1:2],2,as.character)
  jlnet = network(jl2)
#  goid <- sapply(strsplit(names(gset)," "),"[[",1)
#  es <- em[which(em$GO.ID %in% goid),"ES"]
  es <- y[match(names(gset),y$set),"s.dist"]
  colours <- map2color(es,colfunc)
  plot(jlnet, displaylabels = TRUE, label.col = "steelblue",
       edge.lwd = c(jl$edgeSize) * 100,
       arrowhead.cex = 0,
       label.cex = 0.8, vertex.border = "white",vertex.cex = jl$vertexsize,
       vertex.col = colours, edge.col = rgb(0, 0, 0, alpha = 0.5))
}

```


Load data.

```{r,load}

load("scanalyse1.Rdata")

```

## Normalise data

```{r,norm1}

cellmetadata <- data.frame(colnames(comb) ,sapply(strsplit(colnames(comb)," "),"[[",1))
colnames(cellmetadata) <- c("cell","patient")
comb <- CreateSeuratObject(comb, project = "nkmono", assay = "RNA", meta.data = cellmetadata)
comb <- NormalizeData(comb)
comb <- FindVariableFeatures(comb, selection.method = "vst", nfeatures = 2000)
comb <- ScaleData(comb)

```

## PCA and Cluster

```{r,pca1}

comb <- RunPCA(comb, features = VariableFeatures(object = comb))
DimHeatmap(comb, dims = 1, cells = 500, balanced = TRUE)
ElbowPlot(comb)
comb <- JackStraw(comb, num.replicate = 100)
comb <- FindNeighbors(comb, dims = 1:5)
comb <- FindClusters(comb, resolution = 0.2)

```

## UMAP

```{r,umap}

comb <- RunUMAP(comb, dims = 1:8)
DimPlot(comb, reduction = "umap")

```

## Assign cell type with canonical markers

| Cluster ID | Markers | Cell Type |
| --- | --- | --- |
| 0 | IL7R, CCR7 | Naive CD4+ T |
| 1 | CD14, LYZ | CD14+ Mono |
| 2 | IL7R, S100A4 | Memory CD4+ |
| 3 | MS4A1 | B |
| 4 | CD8A | CD8+ T |
| 5 | FCGR3A, MS4A7 | FCGR3A+ Mono |
| 6 | GNLY, NKG7 | NK |
| 7 | FCER1A, CST3 | DC |
| 8 | PPBP | Platelet |

```{r,assign}

message("Naive CD4+ T")
VlnPlot(comb, features = c("IL7R", "CCR7"))

message("CD14+ Mono")
VlnPlot(comb, features = c("CD14", "LYZ"))

message("Memory CD4+ T")
VlnPlot(comb, features = c("IL7R", "S100A4"))

message("B")
VlnPlot(comb, features = c("MS4A1"))

message("CD8+ T")
VlnPlot(comb, features = c("CD8A"))

message("FCGR3A+ Mono")
VlnPlot(comb, features = c("FCGR3A", "MS4A7"))

message("NK")
VlnPlot(comb, features = c("GNLY", "NKG7"))

message("DC")
VlnPlot(comb, features = c("FCER1A", "CST3"))

message("Platelet")
VlnPlot(comb, features = c("PPBP"))

#new.cluster.ids <- c("Naive CD4 T", "CD14+ Mono", "Memory CD4 T", "B", "CD8 T",
#    "FCGR3A+ Mono", "NK", "DC", "Platelet")

FeaturePlot(comb, features = c("MS4A1", "GNLY", "CD3E", "CD14", "FCER1A", "FCGR3A",
  "LYZ", "PPBP", "CD8A"))

#names(new.cluster.ids) <- levels(comb)

```


Now get the marker genes for each cluster.


```{r,markers1}

myvec <- unique(comb[["seurat_clusters"]][,1])

markers <- mclapply(myvec, function(i) { FindMarkers(comb, ident.1 = i ) } , mc.cores=16)

length(markers)

# NK 1 (cluster 0)
head(markers[[1]],10)

# CD14+ monocytes 1 (cluster 1)
head(markers[[2]],10)

# NK 2 (cluster 2)
head(markers[[3]],10)

# CD14+ monocytes 2 (cluster 3)
head(markers[[4]],10)

# FCGR3A+ monocytes (cluster 4)
head(markers[[5]],10)

# cluster 5 unknown - probably T cell
# IL7R = Naive T-cells
# CD3D = CD4+ T cells
# CD3E = T cells
head(markers[[6]],10)

# DC (cluster 6)
head(markers[[7]],10)

# cluster 7 unknown - probably macrophage
# CD14 = macrophages
# S100A12 = macrophages and monocytes
# MS4A6A = macrophages and monocytes
head(markers[[8]],10)

# B cells (cluster 8)
head(markers[[9]],10)

```

As there are two CD14+ mono and NK clusters, it might be a good idea to
look at their differences.

```{r,markers2}

# find all NK markers distinguishing cluster 0 (case) from cluster 2 (ctrl)
nk0.markers <- FindMarkers(comb, ident.1 = 0, ident.2 = 2,only.pos=TRUE)
head(nk0.markers, n = 10)

nk2.markers <- FindMarkers(comb, ident.1 = 2, ident.2 = 0,only.pos=TRUE)
head(nk2.markers, n = 10)

# find all monocyte markers distinguishing cluster 1 (case) from cluster 3 (ctrl)
monocyte1.markers <- FindMarkers(comb, ident.1 = 1, ident.2 = 3, only.pos=TRUE)
head(monocyte1.markers, n = 10)

monocyte3.markers <- FindMarkers(comb, ident.1 = 3, ident.2 = 1, only.pos=TRUE)
head(monocyte3.markers, n = 10)

```

## Assign names to clusters

0. NK 1 (FCGR3A+)

1. CD14+ monocytes

2. NK 2 (IL7R+)

3. CD14+ monocytes (CCL4+)

4. FCGR3A+ monocytes

5. T

6. DC

7. macrophages

8. B

```{r,namingclusters}

new.cluster.ids <- c("NK 1 (FCGRA3A+)", "CD14+ Mono", "NK 2 (IL7R+)", "CD14+ Mono (CCL4+)", "FCGR3A+ Mono", "T",
    "DC", "Macrophage", "B")

names(new.cluster.ids) <- levels(comb)

comb <- RenameIdents(comb, new.cluster.ids)

DimPlot(comb, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend()

```

## count cells by patient

| Sample | Plate | HTO |
| --- | --- | --- |
| CC0003 | 1 | 1 |
| AH0018 | 1 | 3 |
| PM008 | 1 | 4 |
| PM017 | 1 | 5 |
| PM0032 | 2 | 1 |
| PM0028 | 2 | 2 |
| AH0005 | 2 | 3 |
| AH0015 | 2 | 4 |
| PM0027 | 3 | 2 |
| PM0020 | 3 | 3 |
| PM001 | 3 | 4 |
| CC0016 | 3 | 5 |

```{r,count1}

str(Idents(comb))

table(Idents(comb))

cells <- paste(Idents(comb),names(Idents(comb)))

cells2 <- strsplit((stringi::stri_reverse(cells)), " ")
cells2 <- lapply(cells2,function(x) { stringi::stri_reverse(paste(x[2:length(x)],collapse=" "))})
cells2 <- table(unlist(cells2))

cmx <- matrix(cells2,nrow=12)
colnames(cmx) <- unique(sapply(strsplit(gsub(" CC","@",gsub(" AH","@",gsub(" PM","@",names(cells2)))),"@"),"[[",1))

patient_id <- c("AH0005+","AH0015+","AH0018+","CC0003-","CC0016+","PM001+","PM0020-","PM0027-","PM0028-","PM0032-","PM008+","PM017+")
hiv_status <- c(1,1,1,0,1,1,0,0,0,0,1,1)
batch <- factor(c(2,2,1,1,3,3,3,3,2,2,1,1))
pat_df <- data.frame(patient_id,hiv_status,batch) # there were 3 chromium runs contianing multiple samples
pat_df

rownames(cmx) <- patient_id

tcmx <- t(cmx)

tcmx |> kbl(caption="cell counts") |> kable_paper("hover", full_width = F)

ntcmx <- apply(tcmx,2,function(x) { x/sum(x)*100 } )

ntcmx |> kbl(caption="cell proportions") |> kable_paper("hover", full_width = F)

tntcmx <- t(ntcmx)

tntcmx |> kbl(caption="cell proportions") |> kable_paper("hover", full_width = F)

tntcmx_pos <- tntcmx[grep("\\+",rownames(tntcmx)),]

tntcmx_neg <- tntcmx[grep("\\-",rownames(tntcmx)),]

par(mfrow=c(3,3))

lapply(1:ncol(cmx) , function(i) {
  cellname=colnames(tntcmx)[i]
  boxplot(tntcmx_neg[,i],tntcmx_pos[,i],col="white",main=cellname, names=c("HIV-","HIV+"))
  beeswarm(list(tntcmx_neg[,i],tntcmx_pos[,i]),add=TRUE,cex=1.5,col="gray",pch=19)
  tt <- t.test(tntcmx_neg[,i],tntcmx_pos[,i])
  mtext(paste("P=",signif(tt$p.value,3)),cex=0.8)
})

summary(tntcmx_neg)
summary(tntcmx_pos)
ttres <- lapply(1:ncol(cmx) , function(i) {
  t.test(tntcmx_neg[,i],tntcmx_pos[,i])
})
names(ttres) <- colnames(cmx)
ttres

par(mfrow=c(1,1))

```

There were no significant results of the cell counting, however there were some trends that might require further investigation.
In particular relative counts of IL7R+ NK 2 cells were lower, CCL4+ CD14+ Monocytes were lower, while DCs were higher in HIV+
patients.

## Differential expression

We are going to use muscat for pseudobulk analysis.
First need to convert seurat obj to singlecellexperiment object.
Then summarise counts to pseudobulk level.

```{r,pb}

sce <- Seurat::as.SingleCellExperiment(comb, assay = "RNA")

head(colData(sce),2)

colnames(colData(sce))

patient_id <- c("AH0005","AH0015","AH0018","CC0003","CC0016","PM001","PM0020","PM0027","PM0028","PM0032","PM008","PM017")
hiv_status <- c(1,1,1,0,1,1,0,0,0,0,1,1)

colData(sce)$sample_id <- colData(sce)$patient
colData(sce)$cluster_id <- colData(sce)$ident
colData(sce)$hiv <- pat_df[match(colData(sce)$patient,pat_df$patient_id),2]

#muscat library
pb <- aggregateData(sce,
    assay = "counts", fun = "sum",
    by = c("cluster_id", "sample_id"))

# one sheet per subpopulation
assayNames(pb)

t(head(assay(pb)))

plotMDS(assay(pb))

par(mfrow=c(3,3))

lapply(1:length(assays(pb)) , function(i) {
  cellname=names(assays(pb))[i]
  plotMDS(assays(pb)[[i]],cex=1.5,pch=19,col=hiv_status+1,main=paste(cellname))
  mtext("red=HIV+",cex=0.7)
})

par(mfrow=c(1,1))

lapply(1:length(assays(pb)) , function(i) {
  cellname=names(assays(pb))[i]
  plotMDS(assays(pb)[[i]],cex=1.5,pch=19,labels=patient_id,col=hiv_status+1,main=paste("MDS",cellname,"red=HIV+"))
})

plotMDS(assays(pb)[[1]],cex=1.5,labels=patient_id,col=hiv_status+1,main="MDS red=HIV+")

```

Differential expression without batch correction.

```{r,de1}

rownames(pat_df) <- gsub("-","",gsub("\\+","",pat_df[,1]))

deres <- lapply(1:length(assays(pb)), function(i) {
  counts <- assays(pb)[[i]]
  cellname <- names(assays(pb))[i]
  fcounts <- counts[which(rowMeans(counts)>=5),]
  fcounts <- fcounts+1
  dds <- DESeqDataSetFromMatrix(countData = fcounts , colData = pat_df, design = ~ hiv_status)
  res <- DESeq(dds)
  z <- results(res)
  vsd <- vst(dds, blind=FALSE)
  zz<-cbind(as.data.frame(z),assay(vsd),fcounts)
  de <- as.data.frame(zz[order(zz$pvalue),])
  return(de)
})

names(deres) <- names(assays(pb))

nsig1 <- lapply(deres,function(x) {
  nrow(subset(x,padj<0.05))
})

nsig1

make_volcano <- function(dm,name) {
    sig <- subset(dm,padj<0.05)
    N_SIG=nrow(sig)
    N_UP=nrow(subset(sig,log2FoldChange>0))
    N_DN=nrow(subset(sig,log2FoldChange<0))
    HEADER=paste(N_SIG,"@5%FDR,", N_UP, "up", N_DN, "dn")
    plot(dm$log2FoldChange,-log10(dm$pval),cex=0.5,pch=19,col="darkgray",
        main=name, xlab="log FC", ylab="-log10 pval")
    mtext(HEADER)
    grid()
    points(sig$log2FoldChange,-log10(sig$pval),cex=0.5,pch=19,col="red")
}

lapply(1:length(deres),function(i) {
  x <- deres[[i]][,1:6]
  myname <- names(deres)[i]
  make_volcano(dm=x,name=myname)
})

message("Up-regulated")
lapply(deres,function(x) {
  x <- x[,1:6]
  head(subset(x,log2FoldChange>0),10)
})

message("Down-regulated")
lapply(deres,function(x) {
  x <- x[,1:6]
  head(subset(x,log2FoldChange<0),10)
})

if ( ! dir.exists("de_analysis") ) {
  dir.create("de_analysis")
}

lapply(1:length(assays(pb)),function(i) {
  cellname=names(assays(pb))[[i]]
  myres <- deres[[i]]
  filename <- paste(cellname,".tsv",sep="")
  filename <- gsub("\\)","",gsub("\\(","",gsub(" ","_",filename)))
  filename <- paste("de_analysis/",filename,sep="")
  write.table(x=myres,file=filename,sep="\t",quote=FALSE)
})

```

Differential expression with batch correction.

```{r,de2_batch_corrected}

deres <- lapply(1:length(assays(pb)), function(i) {
  counts <- assays(pb)[[i]]
  cellname <- names(assays(pb))[i]
  fcounts <- counts[which(rowMeans(counts)>=5),]
  fcounts <- fcounts+1
  dds <- DESeqDataSetFromMatrix(countData = fcounts , colData = pat_df, design = ~ batch + hiv_status)
  res <- DESeq(dds)
  z <- results(res)
  vsd <- vst(dds, blind=FALSE)
  zz<-cbind(as.data.frame(z),assay(vsd),fcounts)
  de <- as.data.frame(zz[order(zz$pvalue),])
  return(de)
})

names(deres) <- names(assays(pb))

nsig2 <- lapply(deres,function(x) {
  nrow(subset(x,padj<0.05))
})

nsig2

bc <- t(rbind(as.data.frame(nsig1),as.data.frame(nsig2)))
colnames(bc) <- c("uncorrected","corrected")

lapply(1:length(deres),function(i) {
  x <- deres[[i]][,1:6]
  myname <- names(deres)[i]
  make_volcano(dm=x,name=myname)
})

message("Up-regulated")
lapply(deres,function(x) {
  x <- x[,1:6]
  head(subset(x,log2FoldChange>0),10)
})

message("Down-regulated")
lapply(deres,function(x) {
  x <- x[,1:6]
  head(subset(x,log2FoldChange<0),10)
})

if ( ! dir.exists("de_analysis") ) {
  dir.create("de_analysis")
}

lapply(1:length(assays(pb)),function(i) {
  cellname=names(assays(pb))[[i]]
  myres <- deres[[i]]
  filename <- paste(cellname,".tsv",sep="")
  filename <- gsub("\\)","",gsub("\\(","",gsub(" ","_",filename)))
  filename <- paste("de_analysis/",filename,sep="")
  write.table(x=myres,file=filename,sep="\t",quote=FALSE)
})

```

## Pathway analysis

Using mitch package with gene ontology terms.

```{r,mitch1}

#go <- gmt_import("c5.go.v2023.2.Hs.symbols.gmt")
go <- gmt_import("go_2024-11.gmt")
names(go) <- gsub("_"," ",names(go))

str(deres)

if ( ! dir.exists("enrichment_analysis") ) {
  dir.create("enrichment_analysis")
}

# mitch multi
m0 <- mitch_import(deres,DEtype="deseq2",joinType="full")
summary(m0)
dim(m0)
# remove genes with 7/9 or more NA values
m0 <- m0[which(apply(m0,1,function(x) { length(which(is.na(x)))<7  } )),]
dim(m0)
mres0 <- mitch_calc(m0,genesets=go,minsetsize=5,cores=16,priority="effect")

head(mres0$enrichment_result,20) |> kbl(caption="top enrichments") |> kable_paper("hover", full_width = F)

head(subset(mres0$enrichment_result,p.adjustMANOVA<0.01),20) |> kbl(caption="top enrichments") |> kable_paper("hover", full_width = F)

sig <- head(subset(mres0$enrichment_result,p.adjustMANOVA<0.01),30)

sig <- sig[,c(1,4:12)]
rownames(sig) <- sig[,1]
sig[,1]=NULL

colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(as.matrix(sig),trace="none",scale="none",
  margin=c(10,22),col=colfunc(25),cexCol=0.8, cexRow=0.7)

pdf("multimitch_heatmap.pdf")
heatmap.2(as.matrix(sig),trace="none",scale="none",
  margin=c(10,22),col=colfunc(25),cexCol=0.8, cexRow=0.7)
dev.off()

```

mitch single


```{r,mitch2}

par(mfrow=c(1,1))
par(mar=c(5,27,3,1))

message("NK 1 (FCGRA3A+)")
m1 <- mitch_import(deres[[1]],DEtype="deseq2",joinType="full")
mres1 <- mitch_calc(m1,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres1$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="NK 1 (FCGRA3A+)")
  if (! file.exists("enrichment_analysis/NK1.html") ) {
    mitch_report(mres1,outfile="enrichment_analysis/NK1.html")
  }
}

message("CD14+ Mono")
m2 <- mitch_import(deres[[2]],DEtype="deseq2",joinType="full")
mres2 <- mitch_calc(m2,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres2$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="CD14+ Mono")
  if (! file.exists("enrichment_analysis/CD14mono.html") ) {
    mitch_report(mres2,outfile="enrichment_analysis/CD14mono.html")
  }
}

message("NK 2 (IL7R+)")
m3 <- mitch_import(deres[[3]],DEtype="deseq2",joinType="full")
mres3 <- mitch_calc(m3,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres3$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="NK 2 (IL7R+)")
  if (! file.exists("enrichment_analysis/NK2.html") ) {
    mitch_report(mres3,outfile="enrichment_analysis/NK2.html")
  }
}

message("CD14+ Mono (CCL4+)")
m4 <- mitch_import(deres[[4]],DEtype="deseq2",joinType="full")
mres4 <- mitch_calc(m4,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres4$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="CD14+ Mono (CCL4+)")
  if (! file.exists("enrichment_analysis/CD14monoCCL4.html") ) {
    mitch_report(mres4,outfile="enrichment_analysis/CD14monoCCL4.html")
  }
}

message("FCGR3A+ Mono")
m5 <- mitch_import(deres[[5]],DEtype="deseq2",joinType="full")
mres5 <- mitch_calc(m5,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres5$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="FCGR3A+ Mono")
  if (! file.exists("enrichment_analysis/FCGR3Amono.html") ) {
    mitch_report(mres5,outfile="enrichment_analysis/FCGR3Amono.html")
  }
}

message("T")
m6 <- mitch_import(deres[[6]],DEtype="deseq2",joinType="full")
mres6 <- mitch_calc(m6,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres6$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="T")
  if (! file.exists("enrichment_analysis/T.html") ) {
    mitch_report(mres6,outfile="enrichment_analysis/T.html")
  }
}

message("DC")
m7 <- mitch_import(deres[[7]],DEtype="deseq2",joinType="full")
mres7 <- mitch_calc(m7,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres7$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="DC")
  if (! file.exists("enrichment_analysis/DC.html") ) {
    mitch_report(mres7,outfile="enrichment_analysis/DC.html")
  }
}

message("Macrophage")
m8 <- mitch_import(deres[[8]],DEtype="deseq2",joinType="full")
mres8 <- mitch_calc(m8,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres8$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="Macrophage")
  if (! file.exists("enrichment_analysis/macrophage.html") ) {
    mitch_report(mres8,outfile="enrichment_analysis/macrophage.html")
  }
}

message("B")
m9 <- mitch_import(deres[[9]],DEtype="deseq2",joinType="full")
mres9 <- mitch_calc(m9,genesets=go,minsetsize=5,cores=16,priority="effect")
res <- mres9$enrichment_result
res <- subset(res,p.adjustANOVA<0.05)
resup <- subset(res,s.dist>0)
resdn <- subset(res,s.dist<0)
head(resup,10)
head(resdn,10)
s <- c(head(resup$s.dist,10), head(resdn$s.dist,10))
names(s) <- c(head(resup$set,10),head(resdn$set,10))
s <- s[order(s)]
cols <- gsub("1","red",gsub("-1","blue",as.character(sign(s))))
if( length(s) > 1 ) {
  barplot(abs(s),las=1,horiz=TRUE,col=cols,xlab="ES",cex.names=0.8,main="B")
  if (! file.exists("enrichment_analysis/B.html") ) {
    mitch_report(mres9,outfile="enrichment_analysis/B.html")
  }
}

```

Heatmap of combined results

```{r,heat1}

pw <- unique(c(head(subset(mres1$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),5)$set,
  head(subset(mres2$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres3$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres4$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres5$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres6$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres7$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres8$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres9$enrichment_result,p.adjustANOVA<0.01 & s.dist>0.3),6)$set,
  head(subset(mres1$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres2$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres3$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres4$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres5$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres6$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres7$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres8$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set,
  head(subset(mres9$enrichment_result,p.adjustANOVA<0.01 & s.dist< -0.3),6)$set))

x1 <- mres1$enrichment_result[which(mres1$enrichment_result$set %in% pw),c("set","s.dist")]
x2 <- mres2$enrichment_result[which(mres2$enrichment_result$set %in% pw),c("set","s.dist")]
x3 <- mres3$enrichment_result[which(mres3$enrichment_result$set %in% pw),c("set","s.dist")]
x4 <- mres4$enrichment_result[which(mres4$enrichment_result$set %in% pw),c("set","s.dist")]
x5 <- mres5$enrichment_result[which(mres5$enrichment_result$set %in% pw),c("set","s.dist")]
x6 <- mres6$enrichment_result[which(mres6$enrichment_result$set %in% pw),c("set","s.dist")]
x7 <- mres7$enrichment_result[which(mres7$enrichment_result$set %in% pw),c("set","s.dist")]
x8 <- mres8$enrichment_result[which(mres8$enrichment_result$set %in% pw),c("set","s.dist")]
x9 <- mres9$enrichment_result[which(mres9$enrichment_result$set %in% pw),c("set","s.dist")]

jlist <- list(x1,x2,x3,x4,x5,x6,x7,x8,x9)

jj <- join_all(jlist,by="set")
rownames(jj) <- jj$set
jj$set=NULL
colnames(jj) <- names(deres)
colfunc <- colorRampPalette(c("blue", "white", "red"))

heatmap.2(as.matrix(jj),trace="none",scale="none",
  margin=c(10,20),col=colfunc(25),cexCol=0.8)

```

## Differential expression2 KIR3DL2 positive cells 

We are going to use muscat for pseudobulk analysis.
First need to convert seurat obj to singlecellexperiment object.
Then summarise counts to pseudobulk level.

Cells to look at:

* DCs
* NK2
* Macrophage

```{r,KIR3DL2}

sc <- assay(sce)
table(sc[which(rownames(sc)=="KIR3DL2"),])

KIR3DL2_2 <- sc[,sc[which(rownames(sc)=="KIR3DL2"),]>1]
KIR3DL2_0 <- sc[,sc[which(rownames(sc)=="KIR3DL2"),]<1]
summary(colSums(KIR3DL2_2))
summary(colSums(KIR3DL2_0))
par(mfrow=c(2,1))
hist(colSums(KIR3DL2_2),breaks=50,xlim=c(0,1e5),main="KIR3DL2+")
hist(colSums(KIR3DL2_0),breaks=50,xlim=c(0,1e5),main="KIR3DL2-")
par(mfrow=c(1,1))

low <- as.numeric(sc[which(rownames(sc)=="KIR3DL2"),]==0) * 1
med <- as.numeric(sc[which(rownames(sc)=="KIR3DL2"),]==1) * 2
high <- as.numeric(sc[which(rownames(sc)=="KIR3DL2"),]>1) * 3
KIR3DL2 <- data.frame(low,med,high)
KIR3DL2$sum <- rowSums(KIR3DL2)
head(KIR3DL2)
table(KIR3DL2$sum)
colData(sce)$KIR3DL2 <- KIR3DL2$sum
# change cluster id names to incorporate KIR3DL2 counts
colData(sce)$cluster_id <- paste(colData(sce)$cluster_id,colData(sce)$KIR3DL2)

sc_low <- sc[,KIR3DL2$sum==1]
sc_med <- sc[,KIR3DL2$sum==2]
sc_hi <- sc[,KIR3DL2$sum==3]

summary(colSums(sc_low))
summary(colSums(sc_med))
summary(colSums(sc_hi))

dim(sc_low)
dim(sc_med)
dim(sc_hi)

colDat_hi <- subset(colData(sce),KIR3DL2==3)
colDat_med <- subset(colData(sce),KIR3DL2==2)
colDat_low <- subset(colData(sce),KIR3DL2==1)

table(colDat_low$ident)/nrow(colDat_low)*100
table(colDat_med$ident)/nrow(colDat_med)*100
table(colDat_hi$ident)/nrow(colDat_hi)*100

message("Focus on NK 1 (FCGRA3A+) now")

sc_low <- sc[,colData(sce)$KIR3DL2==1 & colData(sce)$ident=="NK 1 (FCGRA3A+)"]
sc_med <- sc[,colData(sce)$KIR3DL2==2 & colData(sce)$ident=="NK 1 (FCGRA3A+)"]
sc_hi <- sc[,colData(sce)$KIR3DL2==3 & colData(sce)$ident=="NK 1 (FCGRA3A+)"]

colDat_hi <- subset(colData(sce),KIR3DL2==3 & ident=="NK 1 (FCGRA3A+)")
sc_hi <- SingleCellExperiment(list(counts=sc_hi), colData=colDat_hi)

colDat_med <- subset(colData(sce),KIR3DL2==2 & ident=="NK 1 (FCGRA3A+)")
sc_med <- SingleCellExperiment(list(counts=sc_med), colData=colDat_med)

colDat_low <- subset(colData(sce),KIR3DL2==1 & ident=="NK 1 (FCGRA3A+)")
sc_low <- SingleCellExperiment(list(counts=sc_low), colData=colDat_low)

#muscat aggregation of NK 1 (FCGRA3A+) cells
pbhi <- aggregateData(sc_hi, assay="counts", fun = "sum", by = c("sample_id"))
pbmed <- aggregateData(sc_med, assay="counts", fun = "sum", by = c("sample_id"))
pblow <- aggregateData(sc_low, assay="counts", fun = "sum", by = c("sample_id"))

colnames(pbhi) <- paste(colnames(pbhi),"hi")
colnames(pbmed) <- paste(colnames(pbmed),"med")
colnames(pblow) <- paste(colnames(pblow),"low")

kx <- cbind(assay(pbhi),assay(pbmed),assay(pblow))

dfhi <- pat_df
dfhi$KIR3DL2 <- 3
dfmed <- pat_df
dfmed$KIR3DL2 <- 2
dflow <- pat_df
dflow$KIR3DL2 <- 1

kdf <- rbind(dfhi,dfmed,dflow)
rownames(kdf) <- colnames(kx)

## focus on HIV negative
kdf1 <- subset(kdf,hiv_status==0 & KIR3DL2!=2)
kx1 <- kx[,which(colnames(kx) %in% rownames(kdf1))]
colSums(kx1)
plotMDS(kx1)

kx1f <- kx1[which(rowMeans(kx1)>=5),]
kx1f <- kx1f+1
dds <- DESeqDataSetFromMatrix(countData = kx1f , colData = kdf1, design = ~ batch + KIR3DL2)
res <- DESeq(dds)
z <- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd),kx1f)
kx1de <- as.data.frame(zz[order(zz$pvalue),])

kx1de[1:50,1:6] |> kbl(caption="KIR3DL2 associated genes in HIV- cells") |> kable_paper("hover", full_width = F)

write.table(x=kx1de,file="KIR3DL2_de_hivneg.tsv",sep="\t",quote=FALSE)

## select HIV positive
kdf2 <- subset(kdf,hiv_status==1 & KIR3DL2!=2)
kx2 <- kx[,which(colnames(kx) %in% rownames(kdf2))]
colSums(kx2)
plotMDS(kx2)

kx2f <- kx2[which(rowMeans(kx2)>=5),]
kx2f <- kx2f+1
dds <- DESeqDataSetFromMatrix(countData = kx2f , colData = kdf2, design = ~ batch + KIR3DL2)
res <- DESeq(dds)
z <- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd),kx2f)
kx2de <- as.data.frame(zz[order(zz$pvalue),])

kx2de[1:50,1:6] |> kbl(caption="KIR3DL2 associated genes in HIV+ cells") |> kable_paper("hover", full_width = F)

write.table(x=kx2de,file="KIR3DL2_de_hivpos.tsv",sep="\t",quote=FALSE)

```

Now make a model that incorporates HIV+ and HIV- cells.


```{r,KIR3DL3}

kdf
head(kx,2)

kdf3 <- subset(kdf, KIR3DL2!=2)

kx3 <- kx[,which(colnames(kx) %in% rownames(kdf3))]

colSums(kx3)
plotMDS(kx3)

dim(kx3)
kx3f <- kx3[which(rowMeans(kx3)>=5),]
dim(kx3f)

# hiv-
dds <- DESeqDataSetFromMatrix(countData = kx3f , colData = kdf3, design = ~ batch + hiv_status + KIR3DL2)

res <- DESeq(dds)
z <- results(res)
vsd <- vst(dds, blind=FALSE)
zz <- cbind(as.data.frame(z),assay(vsd),kx3f)
kx3de <- as.data.frame(zz[order(zz$pvalue),])

kx3de[1:50,1:6] |> kbl(caption="KIR3DL2 associated genes irrespective of HIV status") |> kable_paper("hover", full_width = F)

```

Pathway analysis.

```{r,kirdl3_pw_neg}

k1 <- mitch_import(kx1de,DEtype="deseq2")
k1m <- mitch_calc(k1,genesets=go,minsetsize=5,cores=16,priority="effect")

head(subset(k1m$enrichment_result, s.dist > 0),20) |> kbl(caption="top enrichments (positive correl)") |> kable_paper("hover", full_width = F)
head(subset(k1m$enrichment_result, s.dist < 0),20) |> kbl(caption="top enrichments (negative correl)") |> kable_paper("hover", full_width = F)

up <- head(subset(k1m$enrichment_result, s.dist > 0 & p.adjustANOVA < 0.05),20)
dn <- head(subset(k1m$enrichment_result, s.dist < 0 & p.adjustANOVA < 0.05),20)
top <- rbind(up,dn)
top <- top[order(top$s.dist),]

par(mar=c(5.1, 25.1, 4.1, 2.1))
gsn <- substr(top$set, start = 1, stop = 70 )
barplot(top$s.dist,horiz=TRUE,las=1,xlab="ES",names.arg=gsn,cex.names=0.6,main="HIV neg")

mitch_plots(k1m,outfile="KIR3DL2_assoc_hivneg.pdf")

```

herenow network diagram

```{r,kirdl3_pwn_et,fig.width=13,fig.height=9}

par(mar = c(5.1, 4.1, 4.1, 2.1))

### Network diagram
x <- kx1de
y <- k1m$enrichment_result

#go <- gmt_import("go_2024-11.gmt")
goids <- sapply(strsplit(names(go)," "),"[[",1)

# up 20
scores <- x$stat
names(scores) <- rownames(x)
up <- head(subset(y,p.adjustANOVA<0.05 & s.dist > 0),20)

up |> kbl(caption="Top upregulated pathways") |> kable_paper("hover", full_width = F)

up_go <- sapply(strsplit(up[,1]," "),"[[",1)
up_go <- go[which(goids %in% up_go)]
topgs_up <- lapply(1:length(up_go),function(i) {
  gsname <- names(up_go)[i]
  genes <- up_go[[i]]
  gene_scores <- scores[which(names(scores) %in% genes)]
  top_genes <- names(which(gene_scores>2))
  return(top_genes)
})
names(topgs_up) <- names(up_go)
gs2net(topgs_up,y,colorRampPalette(c("pink","darkred"))(n=100))
topgs_up

#dn 20
dn <- head(subset(y,p.adjustANOVA<0.05 & s.dist < 0),20)

dn |> kbl(caption="Top downregulated pathways")|> kable_paper("hover", full_width = F)

dn_go <- sapply(strsplit(dn[,1]," "),"[[",1)
#dn_go <- dn[,1]
dn_go <- go[which(goids %in% dn_go)]
topgs_dn <- lapply(1:length(dn_go),function(i) {
  gsname <- names(dn_go)[i]
  genes <- dn_go[[i]]
  gene_scores <- scores[which(names(scores) %in% genes)]
  top_genes <- names(which(gene_scores< -2))
  return(top_genes)
})
names(topgs_dn) <- names(dn_go)
gs2net(topgs_dn,y,colorRampPalette(c("darkblue","lightblue"))(n=100))
topgs_dn
gset <- c(topgs_up,topgs_dn)

if ( ! file.exists("KIR3DL2_assoc_hivneg.html") ) {
  mitch_report(res=k1m,outfile="KIR3DL2_assoc_hivneg.html")
}

```

Now in HIV+

```{r,kirdl3_pw_pos}

k2 <- mitch_import(kx2de,DEtype="deseq2")
k2m <- mitch_calc(k2,genesets=go,minsetsize=5,cores=16,priority="effect")

head(subset(k2m$enrichment_result, s.dist > 0),20) |> kbl(caption="top enrichments (positive correl)")  |> kable_paper("hover", full_width = F)
head(subset(k2m$enrichment_result, s.dist < 0),20) |> kbl(caption="top enrichments (negative correl)") |> kable_paper("hover", full_width = F)

up <- head(subset(k2m$enrichment_result, s.dist > 0 & p.adjustANOVA < 0.05),20)
dn <- head(subset(k2m$enrichment_result, s.dist < 0 & p.adjustANOVA < 0.05),20)
top <- rbind(up,dn)
top <- top[order(top$s.dist),]

par(mar=c(5.1, 25.1, 4.1, 2.1))
gsn <- substr(top$set, start = 1, stop = 70 )
barplot(top$s.dist,horiz=TRUE,las=1,xlab="ES",names.arg=gsn,cex.names=0.6,main="HIV pos")

mitch_plots(k2m,outfile="KIR3DL2_assoc_hivpos.pdf")

if ( ! file.exists("KIR3DL2_assoc_hivpos.html") ) {
  mitch_report(res=k2m,outfile="KIR3DL2_assoc_hivpos.html")
}

par(mar=c(5.1, 4.1, 4.1, 2.1))

```

Network diagram

```{r,kirdl3_pw_pos_net,fig.width=13,fig.height=9}

### Network diagram
x <- kx2de
y <- k2m$enrichment_result

# up 20
scores <- x$stat
names(scores) <- rownames(x)
up <- head(subset(y,p.adjustANOVA<0.05 & s.dist > 0),20)

up |> kbl(caption="Top upregulated pathways") |> kable_paper("hover", full_width = F)

up_go <- sapply(strsplit(up[,1]," "),"[[",1)
up_go <- go[which(goids %in% up_go)]
topgs_up <- lapply(1:length(up_go),function(i) {
  gsname <- names(up_go)[i]
  genes <- up_go[[i]]
  gene_scores <- scores[which(names(scores) %in% genes)]
  top_genes <- names(which(gene_scores>2))
  return(top_genes)
})
names(topgs_up) <- names(up_go)
gs2net(topgs_up,y,colorRampPalette(c("pink","darkred"))(n=100))
topgs_up

#dn 20
dn <- head(subset(y,p.adjustANOVA<0.05 & s.dist < 0),20)

dn |> kbl(caption="Top downregulated pathways")|> kable_paper("hover", full_width = F)

dn_go <- sapply(strsplit(dn[,1]," "),"[[",1)
#dn_go <- dn[,1]
dn_go <- go[which(goids %in% dn_go)]
topgs_dn <- lapply(1:length(dn_go),function(i) {
  gsname <- names(dn_go)[i]
  genes <- dn_go[[i]]
  gene_scores <- scores[which(names(scores) %in% genes)]
  top_genes <- names(which(gene_scores< -2))
  return(top_genes)
})
names(topgs_dn) <- names(dn_go)
gs2net(topgs_dn,y,colorRampPalette(c("darkblue","lightblue"))(n=100))
topgs_dn
gset <- c(topgs_up,topgs_dn)

```

Multi-contrast enrichment

```{r,kirdl3_pw2d}

km <- merge(k1,k2,by=0)
rownames(km) <- km[,1]
km[,1]=NULL
colnames(km) <- c("neg","pos")
dim(km)

kmm <- mitch_calc(km,genesets=go,minsetsize=5,cores=16,priority="effect")

head(subset(kmm$enrichment_result, s.dist > 0),20) |> kbl(caption="top multi-enrichment") |> kable_paper("hover", full_width = F)

mitch_plots(kmm,outfile="KIR3DL2_assoc_hivnegpos.pdf")

if ( ! file.exists("KIR3DL2_assoc_hivnegpos.html") ) {
  mitch_report(res=kmm,outfile="KIR3DL2_assoc_hivnegpos.html")
}

```

Pathway analysis irrespective of HIV status.

```{r,kirdl3_pw2}

k3 <- mitch_import(kx3de,DEtype="deseq2")
k3m <- mitch_calc(k3,genesets=go,minsetsize=5,cores=16,priority="effect")

head(subset(k3m$enrichment_result, s.dist > 0),20) |> kbl(caption="top enrichments (positive correl)") |> kable_paper("hover", full_width = F)
head(subset(k3m$enrichment_result, s.dist < 0),20) |> kbl(caption="top enrichments (negative correl)") |> kable_paper("hover", full_width = F)

up <- head(subset(k3m$enrichment_result, s.dist > 0 & p.adjustANOVA < 0.05),20)
dn <- head(subset(k3m$enrichment_result, s.dist < 0 & p.adjustANOVA < 0.05),20)
top <- rbind(up,dn)
top <- top[order(top$s.dist),]

par(mar=c(5.1, 25.1, 4.1, 2.1))
gsn <- substr(top$set, start = 1, stop = 70 )
barplot(top$s.dist,horiz=TRUE,las=1,xlab="ES",names.arg=gsn,cex.names=0.6,main="HIV neg")

mitch_plots(k3m,outfile="KIR3DL2_assoc_all.pdf")

```

## Session information

For reproducibility.

```{r,session}

save.image("scanalyse2.Rdata")

sessionInfo()

```

